File: src/bin/cli/cmd/import.rs
use crate::cli::{ImportFormat, OutputFormat};
use crate::dto::ImportResultDto;
use crate::json_output::{output_error, output_success};
use crate::utils;
use lib::backup::ports::BackupService;
use lib::backup::service::SubRipBackupService;
use lib::import::model::CsvSubtitleRow;
use lib::import::ports::ImportService;
use lib::import::service::{AnchoredImportService, CsvImportService};
use lib::subtitle::model::{Subtitle, SubtitleError, SubtitleText, SubtitleTimestamp};
use lib::subtitle::ports::SubtitleService;
use lib::subtitle::service::SubRipService;
use log::{debug, error, info};
use std::collections::HashMap;
use std::io::{self, Write};
use std::path::Path;

/// Show preview of subtitles to be imported
fn show_preview(subtitles: &[Subtitle]) {
    println!("Preview of {} subtitle(s) to be imported:", subtitles.len());
    println!();

    for (i, subtitle) in subtitles.iter().enumerate() {
        if i >= 5 {
            println!("... and {} more subtitle(s)", subtitles.len() - 5);
            break;
        }
        println!(
            "  [{:>4}] {} --> {}",
            subtitle.index.into_inner(),
            Subtitle::format_timestamp(subtitle.start_time.as_ref()),
            Subtitle::format_timestamp(subtitle.end_time.as_ref())
        );
        let text = subtitle.text.as_str();
        let preview_text = match text.char_indices().nth(50) {
            Some((byte_idx, _)) => format!("{}...", &text[..byte_idx]),
            None => text.to_string(),
        };
        println!("       {}", preview_text.replace('\n', " "));
        println!();
    }
}

/// Ask user for confirmation to proceed with import
fn confirm_import(count: usize) -> anyhow::Result<bool> {
    print!("Proceed with importing {} subtitle(s)? (y/N): ", count);
    io::stdout().flush()?;

    let mut input = String::new();
    io::stdin().read_line(&mut input)?;

    Ok(input.trim().eq_ignore_ascii_case("y"))
}

#[allow(clippy::too_many_arguments)]
pub fn handle(
    srt_file: &str,
    input_file: &str,
    import_format: ImportFormat,
    reference: Option<&str>,
    delimiter: &str,
    dry_run: bool,
    force: bool,
    format: &OutputFormat,
) -> anyhow::Result<()> {
    info!("importing subtitles from {} to {}", input_file, srt_file);

    match import_format {
        ImportFormat::Anchored => {
            if reference.is_none() {
                output_error(
                    format,
                    "missing_reference",
                    "Anchored format requires --reference FILE parameter",
                    Some("Use --reference to specify the SRT file containing timestamps"),
                );
                return Err(anyhow::anyhow!(""));
            }
        }
        ImportFormat::Csv => {
            if delimiter.len() != 1 {
                output_error(
                    format,
                    "invalid_delimiter",
                    &format!("Delimiter must be a single character, got: '{}'", delimiter),
                    None,
                );
                return Err(anyhow::anyhow!(""));
            }
        }
    }

    let delimiter_char = if matches!(import_format, ImportFormat::Csv) {
        delimiter.chars().next().unwrap()
    } else {
        '|'
    };

    let srt_resolved = utils::resolve_existing_path(srt_file)?;
    let input_resolved = utils::resolve_existing_path(input_file)?;

    let subtitles = match import_format {
        ImportFormat::Csv => handle_csv_import(&input_resolved.full_path, delimiter_char, format)?,
        ImportFormat::Anchored => {
            let reference_path = reference.unwrap();
            handle_anchored_import(
                &input_resolved.full_path,
                reference_path,
                &srt_resolved.base_dir,
                format,
            )?
        }
    };

    show_preview(&subtitles);

    if dry_run {
        info!("dry-run mode, operations not executed");

        let min_idx = subtitles
            .iter()
            .map(|s| *s.index.as_ref())
            .min()
            .unwrap_or(0);
        let max_idx = subtitles
            .iter()
            .map(|s| *s.index.as_ref())
            .max()
            .unwrap_or(0);

        let dto = ImportResultDto {
            imported_count: subtitles.len(),
            start_index: min_idx,
            end_index: max_idx,
            total_subtitles: 0, // Unknown in dry-run
            backup_path: "N/A (dry-run)".into(),
            dry_run: true,
        };

        output_success(format, &dto, || {
            println!("Dry-run mode: no subtitles were imported");
        });
        return Ok(());
    }

    if !force {
        if !confirm_import(subtitles.len())? {
            info!("operation cancelled by user");
            println!("Cancelled");
            return Ok(());
        }
        info!("user confirmed import operation");
    }

    let backup_service = SubRipBackupService::new();
    let backup_result = backup_service.create_backup(&srt_resolved.full_path);

    let backup_path = match backup_result {
        Ok(Some(path)) => {
            debug!("backup created: {}", path);
            path
        }
        Ok(None) => {
            debug!("file does not exist, skipping backup");
            "N/A (new file)".to_string()
        }
        Err(e) => {
            error!("failed to create backup: {}", e);
            output_error(
                format,
                "backup_failed",
                &format!("Failed to create backup: {}", e),
                None,
            );
            return Err(anyhow::anyhow!(""));
        }
    };

    let service = SubRipService::new(srt_resolved.base_dir);

    let (imported_count, start_index, end_index, total_subtitles) = match import_format {
        ImportFormat::Csv => {
            debug!("adding {} subtitles...", subtitles.len());
            let mut added_count = 0;
            let mut start_idx = None;
            let mut end_idx = None;

            for subtitle in subtitles {
                match service.add(
                    &srt_resolved.filename,
                    subtitle.start_time,
                    subtitle.end_time,
                    subtitle.text,
                ) {
                    Ok(report) => {
                        if start_idx.is_none() {
                            start_idx = Some(report.new_index);
                        }
                        end_idx = Some(report.new_index);
                        added_count += 1;
                    }
                    Err(e) => {
                        let cli_err = utils::format_subtitle_error(&e, srt_file);
                        output_error(
                            format,
                            &cli_err.code,
                            &cli_err.message,
                            cli_err.hint.as_deref(),
                        );
                        return Err(anyhow::anyhow!(""));
                    }
                }
            }

            let total = match service.get_all(&srt_resolved.filename) {
                Ok(subs) => subs.len(),
                Err(_) => added_count,
            };

            (
                added_count,
                start_idx.unwrap_or(0),
                end_idx.unwrap_or(0),
                total,
            )
        }
        ImportFormat::Anchored => {
            let existing_subs = match service.get_all(&srt_resolved.filename) {
                Ok(subs) => subs,
                Err(SubtitleError::FileNotFound(_)) => Vec::new(),
                Err(e) => {
                    let cli_err = utils::format_subtitle_error(&e, srt_file);
                    output_error(
                        format,
                        &cli_err.code,
                        &cli_err.message,
                        cli_err.hint.as_deref(),
                    );
                    return Err(anyhow::anyhow!(""));
                }
            };

            let mut merged_map: HashMap<u32, Subtitle> = HashMap::new();

            for sub in existing_subs {
                merged_map.insert(*sub.index.as_ref(), sub);
            }

            for sub in &subtitles {
                merged_map.insert(*sub.index.as_ref(), sub.clone());
            }

            let mut merged_subs: Vec<Subtitle> = merged_map.into_values().collect();
            merged_subs.sort_by_key(|s| *s.index.as_ref());

            match service.write_all(&srt_resolved.filename, &merged_subs) {
                Ok(_) => {}
                Err(e) => {
                    let cli_err = utils::format_subtitle_error(&e, srt_file);
                    output_error(
                        format,
                        &cli_err.code,
                        &cli_err.message,
                        cli_err.hint.as_deref(),
                    );
                    return Err(anyhow::anyhow!(""));
                }
            }

            let min_idx = subtitles
                .iter()
                .map(|s| *s.index.as_ref())
                .min()
                .unwrap_or(0);
            let max_idx = subtitles
                .iter()
                .map(|s| *s.index.as_ref())
                .max()
                .unwrap_or(0);

            (subtitles.len(), min_idx, max_idx, merged_subs.len())
        }
    };

    info!(
        "imported {} subtitles successfully (indices {}-{})",
        imported_count, start_index, end_index
    );

    let dto = ImportResultDto {
        imported_count,
        start_index,
        end_index,
        total_subtitles,
        backup_path: backup_path.clone(),
        dry_run: false,
    };

    output_success(format, &dto, || {
        println!("✓ Subtitles imported successfully");
        println!();
        println!("Imported: {} subtitles", imported_count);
        println!("Index range: {}-{}", start_index, end_index);
        println!("Total subtitles: {}", total_subtitles);
        println!("Backup: {}", backup_path);
    });

    Ok(())
}

fn handle_csv_import(
    csv_path: &Path,
    delimiter: char,
    format: &OutputFormat,
) -> anyhow::Result<Vec<Subtitle>> {
    debug!("parsing csv file");
    let import_service = CsvImportService::new();
    let csv_rows = match import_service.parse_csv_file(csv_path, delimiter) {
        Ok(rows) => {
            info!("parsed {} rows from csv file", rows.len());
            rows
        }
        Err(e) => {
            let cli_err = utils::format_subtitle_error(&e, &csv_path.display().to_string());
            output_error(
                format,
                &cli_err.code,
                &cli_err.message,
                cli_err.hint.as_deref(),
            );
            return Err(anyhow::anyhow!(""));
        }
    };

    if csv_rows.is_empty() {
        output_error(format, "empty_csv", "CSV file contains no data rows", None);
        return Err(anyhow::anyhow!(""));
    }

    convert_csv_to_subtitles(csv_rows)
}

fn convert_csv_to_subtitles(csv_rows: Vec<CsvSubtitleRow>) -> anyhow::Result<Vec<Subtitle>> {
    let mut subtitles = Vec::new();

    for csv_row in csv_rows {
        let start_duration = Subtitle::parse_timestamp(&csv_row.start_time).map_err(|e| {
            anyhow::anyhow!(
                "Invalid start timestamp at CSV line {}: {}",
                csv_row.line_number,
                e
            )
        })?;

        let end_duration = Subtitle::parse_timestamp(&csv_row.end_time).map_err(|e| {
            anyhow::anyhow!(
                "Invalid end timestamp at CSV line {}: {}",
                csv_row.line_number,
                e
            )
        })?;

        let start_time = SubtitleTimestamp::try_new(start_duration).map_err(|e| {
            anyhow::anyhow!(
                "Invalid start timestamp at CSV line {}: {}",
                csv_row.line_number,
                e
            )
        })?;

        let end_time = SubtitleTimestamp::try_new(end_duration).map_err(|e| {
            anyhow::anyhow!(
                "Invalid end timestamp at CSV line {}: {}",
                csv_row.line_number,
                e
            )
        })?;

        let text = SubtitleText::try_new(csv_row.text.clone()).map_err(|e| {
            anyhow::anyhow!(
                "Invalid text (empty or whitespace) at CSV line {}: {}",
                csv_row.line_number,
                e
            )
        })?;

        let temp_index = lib::subtitle::model::SubtitleIndex::try_new((subtitles.len() + 1) as u32)
            .expect("valid subtitle index");
        let subtitle = Subtitle::new(temp_index, start_time, end_time, text).map_err(|e| {
            anyhow::anyhow!(
                "Invalid subtitle at CSV line {}: {}",
                csv_row.line_number,
                e
            )
        })?;

        subtitles.push(subtitle);
    }

    Ok(subtitles)
}

fn handle_anchored_import(
    anchored_path: &Path,
    reference_file: &str,
    _base_dir: &Path,
    format: &OutputFormat,
) -> anyhow::Result<Vec<Subtitle>> {
    debug!("parsing anchored format file");

    let import_service = AnchoredImportService::new();
    let anchored_rows = match import_service.parse_anchored_file(anchored_path) {
        Ok(rows) => {
            info!("parsed {} entries from anchored file", rows.len());
            rows
        }
        Err(e) => {
            let cli_err = utils::format_subtitle_error(&e, &anchored_path.display().to_string());
            output_error(
                format,
                &cli_err.code,
                &cli_err.message,
                cli_err.hint.as_deref(),
            );
            return Err(anyhow::anyhow!(""));
        }
    };

    debug!("loading reference file: {}", reference_file);
    let ref_resolved = utils::resolve_existing_path(reference_file)?;
    let service = SubRipService::new(ref_resolved.base_dir);

    let reference_subtitles = match service.get_all(&ref_resolved.filename) {
        Ok(subs) => {
            info!("loaded {} subtitles from reference file", subs.len());
            subs
        }
        Err(e) => {
            let cli_err = utils::format_subtitle_error(&e, reference_file);
            output_error(
                format,
                &cli_err.code,
                &cli_err.message,
                cli_err.hint.as_deref(),
            );
            return Err(anyhow::anyhow!(""));
        }
    };

    let mut ref_map: HashMap<u32, &Subtitle> = HashMap::new();
    for subtitle in &reference_subtitles {
        ref_map.insert(*subtitle.index.as_ref(), subtitle);
    }

    let mut updated_subtitles = Vec::new();

    for anchored_row in &anchored_rows {
        let index = anchored_row.index;

        match ref_map.get(&index) {
            Some(ref_subtitle) => {
                let new_text = SubtitleText::try_new(anchored_row.text.clone()).map_err(|e| {
                    anyhow::anyhow!(
                        "Invalid text (empty or whitespace) at line {}: {}",
                        anchored_row.line_number,
                        e
                    )
                })?;

                let subtitle = Subtitle::new(
                    ref_subtitle.index,
                    ref_subtitle.start_time,
                    ref_subtitle.end_time,
                    new_text,
                )
                .map_err(|e| {
                    anyhow::anyhow!("Failed to create subtitle for index {}: {}", index, e)
                })?;

                updated_subtitles.push(subtitle);
            }
            None => {
                output_error(
                    format,
                    "reference_index_not_found",
                    &format!(
                        "Reference file does not contain subtitle with index {}",
                        index
                    ),
                    Some("Check that the reference file contains all required indices"),
                );
                return Err(anyhow::anyhow!(""));
            }
        }
    }

    updated_subtitles.sort_by_key(|s| *s.index.as_ref());

    Ok(updated_subtitles)
}

File: internal/cli/cmd/import.go
package cmd

import (
	"bufio"
	"errors"
	"fmt"
	"io"
	"log/slog"
	"os"
	"sort"
	"strings"

	"github.com/lebe-dev/submarine/internal/cli"
	"github.com/lebe-dev/submarine/pkg/backup"
	"github.com/lebe-dev/submarine/pkg/importer"
	"github.com/lebe-dev/submarine/pkg/subtitle"
)

// showPreview shows a preview of subtitles to be imported. Port of the Rust
// `show_preview`.
func showPreview(subtitles []subtitle.Subtitle) {
	fmt.Printf("Preview of %d subtitle(s) to be imported:\n", len(subtitles))
	fmt.Println()

	for i := range subtitles {
		if i >= 5 {
			fmt.Printf("... and %d more subtitle(s)\n", len(subtitles)-5)
			break
		}
		s := &subtitles[i]
		fmt.Printf(
			"  [%4d] %s --> %s\n",
			s.Index.Value(),
			subtitle.FormatTimestamp(s.StartTime.Value()),
			subtitle.FormatTimestamp(s.EndTime.Value()),
		)
		text := s.Text.Value()
		previewText := truncateAt50(text)
		fmt.Printf("       %s\n", strings.ReplaceAll(previewText, "\n", " "))
		fmt.Println()
	}
}

// truncateAt50 mirrors the Rust `text.char_indices().nth(50)` truncation: if the
// text has more than 50 characters, truncate to the first 50 characters and
// append "...", otherwise return the text unchanged.
func truncateAt50(text string) string {
	count := 0
	for byteIdx := range text {
		if count == 50 {
			return text[:byteIdx] + "..."
		}
		count++
	}
	return text
}

// confirmImport asks the user for confirmation to proceed with import. Port of
// the Rust `confirm_import`.
func confirmImport(count int) (bool, error) {
	fmt.Printf("Proceed with importing %d subtitle(s)? (y/N): ", count)

	reader := bufio.NewReader(os.Stdin)
	input, err := reader.ReadString('\n')
	if err != nil && !errors.Is(err, io.EOF) {
		return false, err
	}

	return strings.EqualFold(strings.TrimSpace(input), "y"), nil
}

// HandleImport is the port of the Rust `cmd::import::handle`.
func HandleImport(
	srtFile string,
	inputFile string,
	importFormat cli.ImportFormat,
	reference *string,
	delimiter string,
	dryRun bool,
	force bool,
	format cli.OutputFormat,
) error {
	slog.Info(fmt.Sprintf("importing subtitles from %s to %s", inputFile, srtFile))

	switch importFormat {
	case cli.ImportFormatAnchored:
		if reference == nil {
			cli.OutputError(
				format,
				"missing_reference",
				"Anchored format requires --reference FILE parameter",
				importStrPtr("Use --reference to specify the SRT file containing timestamps"),
			)
			return errors.New("")
		}
	case cli.ImportFormatCsv:
		// Rust uses `delimiter.len() != 1`, which is the BYTE length of the str.
		if len(delimiter) != 1 {
			cli.OutputError(
				format,
				"invalid_delimiter",
				fmt.Sprintf("Delimiter must be a single character, got: '%s'", delimiter),
				nil,
			)
			return errors.New("")
		}
	}

	var delimiterChar rune
	if importFormat == cli.ImportFormatCsv {
		// Rust `delimiter.chars().next().unwrap()` takes the first char. Since
		// the byte-length check above guaranteed a single byte, this is the
		// first (and only) rune.
		delimiterChar = []rune(delimiter)[0]
	} else {
		delimiterChar = '|'
	}

	srtResolved, err := cli.ResolveExistingPath(srtFile)
	if err != nil {
		return err
	}
	inputResolved, err := cli.ResolveExistingPath(inputFile)
	if err != nil {
		return err
	}

	var subtitles []subtitle.Subtitle
	switch importFormat {
	case cli.ImportFormatCsv:
		subtitles, err = handleCsvImport(inputResolved.FullPath, delimiterChar, format)
		if err != nil {
			return err
		}
	case cli.ImportFormatAnchored:
		referencePath := *reference
		subtitles, err = handleAnchoredImport(inputResolved.FullPath, referencePath, srtResolved.BaseDir, format)
		if err != nil {
			return err
		}
	}

	showPreview(subtitles)

	if dryRun {
		slog.Info("dry-run mode, operations not executed")

		minIdx := minIndex(subtitles)
		maxIdx := maxIndex(subtitles)

		dto := cli.ImportResultDto{
			ImportedCount:  len(subtitles),
			StartIndex:     minIdx,
			EndIndex:       maxIdx,
			TotalSubtitles: 0, // Unknown in dry-run
			BackupPath:     "N/A (dry-run)",
			DryRun:         true,
		}

		cli.OutputSuccess(format, dto, func() {
			fmt.Println("Dry-run mode: no subtitles were imported")
		})
		return nil
	}

	if !force {
		confirmed, cErr := confirmImport(len(subtitles))
		if cErr != nil {
			return cErr
		}
		if !confirmed {
			slog.Info("operation cancelled by user")
			fmt.Println("Cancelled")
			return nil
		}
		slog.Info("user confirmed import operation")
	}

	backupService := backup.NewSubRipBackupService()
	backupResult, backupErr := backupService.CreateBackup(srtResolved.FullPath)

	var backupPath string
	switch {
	case backupErr != nil:
		slog.Error(fmt.Sprintf("failed to create backup: %s", backupErr))
		cli.OutputError(
			format,
			"backup_failed",
			fmt.Sprintf("Failed to create backup: %s", backupErr),
			nil,
		)
		return errors.New("")
	case backupResult != nil:
		slog.Debug(fmt.Sprintf("backup created: %s", *backupResult))
		backupPath = *backupResult
	default:
		slog.Debug("file does not exist, skipping backup")
		backupPath = "N/A (new file)"
	}

	service := subtitle.NewSubRipService(srtResolved.BaseDir)

	var importedCount int
	var startIndex uint32
	var endIndex uint32
	var totalSubtitles int

	switch importFormat {
	case cli.ImportFormatCsv:
		slog.Debug(fmt.Sprintf("adding %d subtitles...", len(subtitles)))
		addedCount := 0
		var startIdx *uint32
		var endIdx *uint32

		for i := range subtitles {
			report, addErr := service.Add(
				srtResolved.Filename,
				subtitles[i].StartTime,
				subtitles[i].EndTime,
				subtitles[i].Text,
			)
			if addErr != nil {
				var subErr *subtitle.SubtitleError
				if errors.As(addErr, &subErr) {
					cliErr := cli.FormatSubtitleError(subErr, srtFile)
					cli.OutputError(format, cliErr.Code, cliErr.Message, cliErr.Hint)
					return errors.New("")
				}
				return addErr
			}
			if startIdx == nil {
				v := report.NewIndex
				startIdx = &v
			}
			ev := report.NewIndex
			endIdx = &ev
			addedCount++
		}

		total := addedCount
		if subs, getErr := service.GetAll(srtResolved.Filename); getErr == nil {
			total = len(subs)
		}

		importedCount = addedCount
		startIndex = derefOrZero(startIdx)
		endIndex = derefOrZero(endIdx)
		totalSubtitles = total

	case cli.ImportFormatAnchored:
		var existingSubs []subtitle.Subtitle
		subs, getErr := service.GetAll(srtResolved.Filename)
		if getErr != nil {
			var subErr *subtitle.SubtitleError
			if errors.As(getErr, &subErr) && subErr.Kind == subtitle.ErrFileNotFound {
				existingSubs = nil
			} else if errors.As(getErr, &subErr) {
				cliErr := cli.FormatSubtitleError(subErr, srtFile)
				cli.OutputError(format, cliErr.Code, cliErr.Message, cliErr.Hint)
				return errors.New("")
			} else {
				return getErr
			}
		} else {
			existingSubs = subs
		}

		mergedMap := make(map[uint32]subtitle.Subtitle)

		for i := range existingSubs {
			mergedMap[existingSubs[i].Index.Value()] = existingSubs[i]
		}

		for i := range subtitles {
			mergedMap[subtitles[i].Index.Value()] = subtitles[i]
		}

		mergedSubs := make([]subtitle.Subtitle, 0, len(mergedMap))
		for _, sub := range mergedMap {
			mergedSubs = append(mergedSubs, sub)
		}
		sort.SliceStable(mergedSubs, func(a, b int) bool {
			return mergedSubs[a].Index.Value() < mergedSubs[b].Index.Value()
		})

		if writeErr := service.WriteAll(srtResolved.Filename, mergedSubs); writeErr != nil {
			var subErr *subtitle.SubtitleError
			if errors.As(writeErr, &subErr) {
				cliErr := cli.FormatSubtitleError(subErr, srtFile)
				cli.OutputError(format, cliErr.Code, cliErr.Message, cliErr.Hint)
				return errors.New("")
			}
			return writeErr
		}

		minIdx := minIndex(subtitles)
		maxIdx := maxIndex(subtitles)

		importedCount = len(subtitles)
		startIndex = minIdx
		endIndex = maxIdx
		totalSubtitles = len(mergedSubs)
	}

	slog.Info(fmt.Sprintf(
		"imported %d subtitles successfully (indices %d-%d)",
		importedCount, startIndex, endIndex,
	))

	dto := cli.ImportResultDto{
		ImportedCount:  importedCount,
		StartIndex:     startIndex,
		EndIndex:       endIndex,
		TotalSubtitles: totalSubtitles,
		BackupPath:     backupPath,
		DryRun:         false,
	}

	cli.OutputSuccess(format, dto, func() {
		fmt.Println("✓ Subtitles imported successfully")
		fmt.Println()
		fmt.Printf("Imported: %d subtitles\n", importedCount)
		fmt.Printf("Index range: %d-%d\n", startIndex, endIndex)
		fmt.Printf("Total subtitles: %d\n", totalSubtitles)
		fmt.Printf("Backup: %s\n", backupPath)
	})

	return nil
}

// handleCsvImport parses a CSV file and converts it to subtitles. Port of the
// Rust `handle_csv_import`.
func handleCsvImport(csvPath string, delimiter rune, format cli.OutputFormat) ([]subtitle.Subtitle, error) {
	slog.Debug("parsing csv file")
	importService := importer.NewCsvImportService()
	csvRows, err := importService.ParseCsvFile(csvPath, delimiter)
	if err != nil {
		var subErr *subtitle.SubtitleError
		if errors.As(err, &subErr) {
			cliErr := cli.FormatSubtitleError(subErr, csvPath)
			cli.OutputError(format, cliErr.Code, cliErr.Message, cliErr.Hint)
			return nil, errors.New("")
		}
		return nil, err
	}
	slog.Info(fmt.Sprintf("parsed %d rows from csv file", len(csvRows)))

	if len(csvRows) == 0 {
		cli.OutputError(format, "empty_csv", "CSV file contains no data rows", nil)
		return nil, errors.New("")
	}

	return convertCsvToSubtitles(csvRows)
}

// convertCsvToSubtitles converts parsed CSV rows into validated subtitles. Port
// of the Rust `convert_csv_to_subtitles`.
func convertCsvToSubtitles(csvRows []importer.CsvSubtitleRow) ([]subtitle.Subtitle, error) {
	var subtitles []subtitle.Subtitle

	for _, csvRow := range csvRows {
		startDuration, err := subtitle.ParseTimestamp(csvRow.StartTime)
		if err != nil {
			return nil, fmt.Errorf("Invalid start timestamp at CSV line %d: %s", csvRow.LineNumber, err)
		}

		endDuration, err := subtitle.ParseTimestamp(csvRow.EndTime)
		if err != nil {
			return nil, fmt.Errorf("Invalid end timestamp at CSV line %d: %s", csvRow.LineNumber, err)
		}

		startTime, err := subtitle.NewSubtitleTimestamp(startDuration)
		if err != nil {
			return nil, fmt.Errorf("Invalid start timestamp at CSV line %d: %s", csvRow.LineNumber, err)
		}

		endTime, err := subtitle.NewSubtitleTimestamp(endDuration)
		if err != nil {
			return nil, fmt.Errorf("Invalid end timestamp at CSV line %d: %s", csvRow.LineNumber, err)
		}

		text, err := subtitle.NewSubtitleText(csvRow.Text)
		if err != nil {
			return nil, fmt.Errorf("Invalid text (empty or whitespace) at CSV line %d: %s", csvRow.LineNumber, err)
		}

		tempIndex, idxErr := subtitle.NewSubtitleIndex(uint32(len(subtitles) + 1))
		if idxErr != nil {
			panic("valid subtitle index")
		}
		sub, err := subtitle.NewSubtitle(tempIndex, startTime, endTime, text)
		if err != nil {
			return nil, fmt.Errorf("Invalid subtitle at CSV line %d: %s", csvRow.LineNumber, err)
		}

		subtitles = append(subtitles, sub)
	}

	return subtitles, nil
}

// handleAnchoredImport parses an anchored format file against a reference SRT
// file. Port of the Rust `handle_anchored_import`.
func handleAnchoredImport(anchoredPath string, referenceFile string, baseDir string, format cli.OutputFormat) ([]subtitle.Subtitle, error) {
	_ = baseDir
	slog.Debug("parsing anchored format file")

	importService := importer.NewAnchoredImportService()
	anchoredRows, err := importService.ParseAnchoredFile(anchoredPath)
	if err != nil {
		var subErr *subtitle.SubtitleError
		if errors.As(err, &subErr) {
			cliErr := cli.FormatSubtitleError(subErr, anchoredPath)
			cli.OutputError(format, cliErr.Code, cliErr.Message, cliErr.Hint)
			return nil, errors.New("")
		}
		return nil, err
	}
	slog.Info(fmt.Sprintf("parsed %d entries from anchored file", len(anchoredRows)))

	slog.Debug(fmt.Sprintf("loading reference file: %s", referenceFile))
	refResolved, err := cli.ResolveExistingPath(referenceFile)
	if err != nil {
		return nil, err
	}
	service := subtitle.NewSubRipService(refResolved.BaseDir)

	referenceSubtitles, err := service.GetAll(refResolved.Filename)
	if err != nil {
		var subErr *subtitle.SubtitleError
		if errors.As(err, &subErr) {
			cliErr := cli.FormatSubtitleError(subErr, referenceFile)
			cli.OutputError(format, cliErr.Code, cliErr.Message, cliErr.Hint)
			return nil, errors.New("")
		}
		return nil, err
	}
	slog.Info(fmt.Sprintf("loaded %d subtitles from reference file", len(referenceSubtitles)))

	refMap := make(map[uint32]*subtitle.Subtitle)
	for i := range referenceSubtitles {
		refMap[referenceSubtitles[i].Index.Value()] = &referenceSubtitles[i]
	}

	var updatedSubtitles []subtitle.Subtitle

	for i := range anchoredRows {
		anchoredRow := &anchoredRows[i]
		index := anchoredRow.Index

		refSubtitle, ok := refMap[index]
		if !ok {
			cli.OutputError(
				format,
				"reference_index_not_found",
				fmt.Sprintf("Reference file does not contain subtitle with index %d", index),
				importStrPtr("Check that the reference file contains all required indices"),
			)
			return nil, errors.New("")
		}

		newText, textErr := subtitle.NewSubtitleText(anchoredRow.Text)
		if textErr != nil {
			return nil, fmt.Errorf("Invalid text (empty or whitespace) at line %d: %s", anchoredRow.LineNumber, textErr)
		}

		sub, subErr := subtitle.NewSubtitle(
			refSubtitle.Index,
			refSubtitle.StartTime,
			refSubtitle.EndTime,
			newText,
		)
		if subErr != nil {
			return nil, fmt.Errorf("Failed to create subtitle for index %d: %s", index, subErr)
		}

		updatedSubtitles = append(updatedSubtitles, sub)
	}

	sort.SliceStable(updatedSubtitles, func(a, b int) bool {
		return updatedSubtitles[a].Index.Value() < updatedSubtitles[b].Index.Value()
	})

	return updatedSubtitles, nil
}

// minIndex returns the minimum subtitle index, or 0 when the slice is empty.
// Mirrors the Rust `.map(|s| *s.index.as_ref()).min().unwrap_or(0)`.
func minIndex(subtitles []subtitle.Subtitle) uint32 {
	if len(subtitles) == 0 {
		return 0
	}
	m := subtitles[0].Index.Value()
	for i := range subtitles {
		v := subtitles[i].Index.Value()
		if v < m {
			m = v
		}
	}
	return m
}

// maxIndex returns the maximum subtitle index, or 0 when the slice is empty.
// Mirrors the Rust `.map(|s| *s.index.as_ref()).max().unwrap_or(0)`.
func maxIndex(subtitles []subtitle.Subtitle) uint32 {
	if len(subtitles) == 0 {
		return 0
	}
	m := subtitles[0].Index.Value()
	for i := range subtitles {
		v := subtitles[i].Index.Value()
		if v > m {
			m = v
		}
	}
	return m
}

// derefOrZero returns *p or 0 when p is nil. Mirrors the Rust
// `Option::unwrap_or(0)`.
func derefOrZero(p *uint32) uint32 {
	if p == nil {
		return 0
	}
	return *p
}

// importStrPtr returns a pointer to s (helper for optional hints).
func importStrPtr(s string) *string { return &s }
